### PR TITLE
New Code Coverage script and buildRpm script

### PR DESCRIPTION
**Pull request by craig-cogdill**
The new way code coverage for C++ repos (excluding Research) will work is that this rpm will install the code coverage executable in `/usr/local/probe/bin/CodeCoverage.py`. Simply run the script from any C++ repo that uses CMake.

### CODE REVIEW

File: scripts/CodeCoverage.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    ARG1_FLAGS = '-DCMAKE_CXX_COMPILER_ARG1:STRING=\'-Wall -Werror -g -gdwarf-2 -fno-elide-constructors -fprofile-arcs -ftest-coverage -O0 -fPIC -m64 -Wl,-rpath -Wl,. -Wl,-rpath -Wl,/usr/local/probe/lib -Wl,-rpath -Wl,/usr/local/probe/lib64 -fno-inline -fno-inline-small-functions -fno-default-inline\''
+    COMPILER_EXE_FLAG = '-DCMAKE_CXX_COMPILER=/usr/local/probe/bin/g++'
+    CMAKELISTS_DIR = '..'
+    CMAKE_STR = CMAKE + ' ' + DEBUG_FLAG + ' ' + VERSION_FLAG + ' ' + ARG1_FLAGS + ' ' + COMPILER_EXE_FLAG + ' ' + CMAKELISTS_DIR
+    CMAKE_CMD = [CMAKE_STR]
+    ret_code = subprocess.check_call(CMAKE_CMD, stderr=subprocess.STDOUT, shell=True)
+    print "CMake return code: " + str(ret_code)
+
+def compile_project():
+    COMPILE_CMD = ['make -j']
```
Comment (craig-cogdill) [inline scripts/CodeCoverage.py]: Change to -j8

File: scripts/CodeCoverage.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+def run_UnitTestRunner(launch_dir):
+    CP_RUNNER_SCRIPT = 'cp ' + launch_dir + '/scripts/unitTestRunner.sh ' + launch_dir + '/build'
+    CP_RUNNER_SCRIPT_CMD = [CP_RUNNER_SCRIPT]
+    ret_code = subprocess.check_call(CP_RUNNER_SCRIPT_CMD, stderr=subprocess.STDOUT, shell=True)
+    print "Copy script return code: " + str(ret_code)
+    RUNNER_SCRIPT_CMD = ['sh unitTestRunner.sh']
+    try:
+        ret_code = subprocess.check_call(RUNNER_SCRIPT_CMD, stderr=subprocess.STDOUT, shell=True)
+    except:
+        print "UnitTestRunner return code: bad"
```
Comment (craig-cogdill) [inline scripts/CodeCoverage.py]: Use return code somewhere? Fail when there is a failure vs continuing

File: scripts/CodeCoverage.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    formatted_list = None
+    if user_list is not None:
+        for whitelist_item in user_list.split():
+            print "Split into : " + whitelist_item
+            if formatted_list is None:
+                formatted_list = '.*' + whitelist_item + '$'
+            else:
+                formatted_list = formatted_list + '|.*'+whitelist_item+'$'
+    else:
+        print "list is empty" #Write something better here
```
Comment (craig-cogdill) [inline scripts/CodeCoverage.py]: Forgot to actually do the comment

**Comment by craig-cogdill** [comment]
from Kjell: 

To the reviewers, code coverage will be run like this example:

```
cd FileIO
/usr/local/probe/bin/CodeCoverage.py
```

You can view it in a browser:

```
firefox coverage/coverage_FileIO.html
```

**Comment by craig-cogdill** [comment]
👍 from Kjell

File: scripts/CodeCoverage.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import sys
+import argparse
+import shutil
+import zipfile
+import ntpath
+import subprocess
+
+DEFAULT_WHITELIST = ".*cpp$"
+DEFAULT_BLACKLIST = "/usr/local/probe/.*"
+
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:14]: Why the linefeed? (just curious)

File: scripts/CodeCoverage.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+CMAKEFILEPATH = "CMakeLists.txt"
+LIBRARY_TO_BUILD_REGEX = 'SET\(LIBRARY_TO_BUILD (.*)\)$'
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:16]: Inconsistent string quotation mark -- " or ' should be used for all strings, escape as needed within. :)

File: scripts/CodeCoverage.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+def get_file_content(filename):
+    with open(filename) as f:
+        content = f.readlines()
+        return content
+
+def get_library_to_build_name(line_str):
+    match_obj = re.search(LIBRARY_TO_BUILD_REGEX, line_str, re.M|re.I)
+    if match_obj:
+        return True, match_obj.group(1)
+    return False, ""
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:27]: Nitpick, can be ignored -- someone might not immediately realize you're returning a tuple here ;) 

return (False, "")

File: scripts/CodeCoverage.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+def get_project_name_from_CMakeLists_file(full_file_path):
+    for line in get_file_content(filename=full_file_path):
+        found, project_name = get_library_to_build_name(line_str=line)
+        if found:
+            return True, project_name
+    return False, ""
+
+def print_error_and_usage(argParser, error):
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:36]: Should be consistent with either snake_case or camelHump

File: scripts/CodeCoverage.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    print "Error:  " + error + "\n"
+    print argParser.print_help()
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:38]: - call print as a function to be compatible with "modern" python 
- consider adjusting your string concatenation, the following is the standard (both examples in one)

```
In [3]: print("Error: {}\n".format("blah"))
Error: blah
```

File: scripts/CodeCoverage.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    sys.exit(2)
+
+def santize_input_args(arg_parser, args):
+    if len(sys.argv) > 3:
+        print_error_and_usage(argParser=arg_parser, error="Too many arguments supplied.")
+
+def destroy_directory(dir_path):
+    shutil.rmtree(dir_path)
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:46]: Redundant function -- shutil.rmtree(dir_path) does the same thing ;)

File: scripts/CodeCoverage.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+def clean_and_build_directory(dir_path):
+    if os.path.exists(dir_path):
+        destroy_directory(dir_path=dir_path)
+    os.makedirs(dir_path)
+
+def remove_file_extension(filename):
+    return os.path.splitext(filename)[0]
+
+def strip_file_from_full_path(full_file_path):
+    return ntpath.basename(full_file_path)
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:57]: Any need for wrapper function?

File: scripts/CodeCoverage.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+def destroy_directory(dir_path):
+    shutil.rmtree(dir_path)
+
+def clean_and_build_directory(dir_path):
+    if os.path.exists(dir_path):
+        destroy_directory(dir_path=dir_path)
+    os.makedirs(dir_path)
+
+def remove_file_extension(filename):
+    return os.path.splitext(filename)[0]
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:54]: Same here, unless you need a wrapper function for anything, they can probably be removed and you can just called os.path.whatever

File: scripts/CodeCoverage.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+def run_CMake_cmd():
+    CMAKE = '/usr/local/probe/bin/cmake'
+    DEBUG_FLAG = '-DUSE_LR_DEBUG=ON'
+    VERSION_FLAG = '-DVERSION=' + str(1)
+    ARG1_FLAGS = '-DCMAKE_CXX_COMPILER_ARG1:STRING=\'-Wall -Werror -g -gdwarf-2 -fno-elide-constructors -fprofile-arcs -ftest-coverage -O0 -fPIC -m64 -Wl,-rpath -Wl,. -Wl,-rpath -Wl,/usr/local/probe/lib -Wl,-rpath -Wl,/usr/local/probe/lib64 -fno-inline -fno-inline-small-functions -fno-default-inline\''
+    COMPILER_EXE_FLAG = '-DCMAKE_CXX_COMPILER=/usr/local/probe/bin/g++'
+    CMAKELISTS_DIR = '..'
+    CMAKE_STR = CMAKE + ' ' + DEBUG_FLAG + ' ' + VERSION_FLAG + ' ' + ARG1_FLAGS + ' ' + COMPILER_EXE_FLAG + ' ' + CMAKELISTS_DIR
+    CMAKE_CMD = [CMAKE_STR]
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:76]: I personally would refactor this concentration to use format, and more line feeds.. something like..

```
CMAKE_STR = "{0} {1} {2} {3}".format(
   CMAKE,
   DEBUG_FLAG,
   VERSION_FLAG,
    ....
)
```

And so on. It's easier to organize your whitespace and understand for a maintainer

File: scripts/CodeCoverage.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+def run_UnitTestRunner(launch_dir):
+    CP_RUNNER_SCRIPT = 'cp ' + launch_dir + '/scripts/unitTestRunner.sh ' + launch_dir + '/build'
+    CP_RUNNER_SCRIPT_CMD = [CP_RUNNER_SCRIPT]
+    ret_code = subprocess.check_call(CP_RUNNER_SCRIPT_CMD, stderr=subprocess.STDOUT, shell=True)
+    print "Copy script return code: " + str(ret_code)
+    RUNNER_SCRIPT_CMD = ['sh unitTestRunner.sh']
+    try:
+        ret_code = subprocess.check_call(RUNNER_SCRIPT_CMD, stderr=subprocess.STDOUT, shell=True)
+    except:
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:102]: You can be specific about your exceptions, check into what you expect it to throw.

```
except Exception as e:
    print("exception bro: {0}".format(e))
```

Maybe like..

```
In [3]: import subprocess

In [4]: try:
   ...:     p = subprocess.Popen('ls fjska')
   ...: except OSError as ose:
   ...:     print(ose)
   ...:     
[Errno 2] No such file or directory
```

File: scripts/CodeCoverage.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    unzipped_dir_name = remove_file_extension(filename=strip_file_from_full_path(full_file_path=full_file_path))
+    full_unzipped_dir_path = directory_to_extract_to + "/" + unzipped_dir_name
+    if os.path.exists(full_unzipped_dir_path):
+        destroy_directory(full_unzipped_dir_path)
+    to_unzip.extractall(directory_to_extract_to)
+    to_unzip.close()
+
+
+def run_CMake_cmd():
+    CMAKE = '/usr/local/probe/bin/cmake'
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:69]: use consistent naming -- no need for CONSTANT style naming on strings in python.
also either ' or " to quote strings, not a mix

```
cmake = 'foo'
```

File: scripts/CodeCoverage.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    SORT_PERCENTAGE = '--sort-percentage'
+    FILTER = '--filter=\"'+whitelist_filter+'\"'
+    EXCLUDE = '--exclude=\"'+blacklist_filter+'\"'
+    GCOV_EXE = '--gcov-executable /usr/local/probe/bin/gcov'
+    EXCLUDE_UNREACHABLE = '--exclude-unreachable-branches'
+    HTML_FLAGS = '--html --html-details'   
+    OUTPUT_FILE = '-o coverage_' + project_name + '.html'
+    FLAGS = VERBOSE + ' ' + SORT_PERCENTAGE + ' ' + FILTER + ' ' + EXCLUDE + ' ' + GCOV_EXE + ' ' + EXCLUDE_UNREACHABLE + ' ' + HTML_FLAGS
+    GCOVR_CMD_STR = GCOVR + ' ' + FLAGS + ' '  + OUTPUT_FILE
+    try:
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:118]: As above, consider changing this to the format style concatenation

File: scripts/CodeCoverage.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        ret_code = subprocess.check_call([GCOVR_CMD_STR], stderr=subprocess.STDOUT, shell=True)
+    except:
+        print "Gcovr return code: bad"
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py]:
```
In [3]: import subprocess

In [4]: try:
   ...:     p = subprocess.Popen('ls fjska')
   ...: except OSError as ose:
   ...:     print(ose)
   ...:     
[Errno 2] No such file or directory
```

File: scripts/CodeCoverage.py [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+def generate_gcovr_filter(formatted_user_list, default_list):
+    if formatted_user_list is None:
+        return default_list
+    else:
+        return default_list + '|' + formatted_user_list
+
+def main(argv):
+    argParser = argparse.ArgumentParser(description="Generate code coverage information for a Network Monitor repository that uses CMake")
+    argParser.add_argument("-w", "--whitelist", dest="whitelist", metavar="HEADER_FILE", help="Whitelist a header or template file for code coverage")
+    argParser.add_argument("-b", "--blacklist", dest="blacklist", metavar="SOURCE_FILE", help="Blacklist a source file for code coverage")
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py]: Too long, break em up!

```
argParser.add_argument(
    "-w", 
    "--whitelist",
    dest="whitelist",
    metavar="HEADER_FILE",
    help="Whitelist a header or template file for code coverage")
```

much more readable, no?

File: scripts/CodeCoverage.py [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+    clean_and_build_directory(dir_path="build")
+
+    os.chdir(LAUNCH_DIR + '/build')
+    run_CMake_cmd()
+    compile_project()
+    run_UnitTestRunner(launch_dir=LAUNCH_DIR)
+    os.chdir(LAUNCH_DIR + '/coverage')
+
+    gcovr_whitelist = generate_gcovr_filter(formatted_user_list=format_user_list(user_list=USER_WHITELIST),
+                                            default_list=DEFAULT_WHITELIST)
```
Comment (mblhaunted) [inline scripts/CodeCoverage.py:199]: nothing wrong here, my preference is to give them all a line, but its up to you:

```
gcovr_whitelist = generate_gcovr_filter(
    formatted_user_list=format_user_list(user_list=USER_WHITELIST),
    default_list=DEFAULT_WHITELIST)
```

File: packaging/CodeCoverage.spec [hunk 19]
```diff
@@ -0,0 +1,35 @@
+Name:          CodeCoverage
+Version:       %{version}
+Release:       1%{?dist}
+Summary:       A wrapper for building and running code coverage on NetMon C++ repositories
+Group:         Development/Tools
+License:       MIT
```
Comment (john-gress) [inline packaging/CodeCoverage.spec:6]: LogRhythm Copyright. Look at Research.

File: scripts/buildRpm.sh [hunk 20]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -e
+
+PACKAGE=CodeCoverage
+VERSION=1.0
+
+PWD=`pwd`
+CWD=$PWD/$PACKAGE
+DISTDIR=$CWD/dist/$PACKAGE
+PATH=$PATH:/usr/local/probe/bin:$PATH
```
Comment (john-gress) [inline scripts/buildRpm.sh]: This needs to be cleaned up. PATH probably doesn't need to be set here and I don't think DISTDIR is needed.

File: scripts/buildRpm.sh [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+# As version number we use the commit number on HEAD 
+# we do not bother with other branches for now
+GIT_VERSION=`git rev-list --branches HEAD | wc -l`
+VERSION="1.$GIT_VERSION"
+
+echo "Building $PACKAGE, type: $BUILD_TYPE, version: $VERSION"
+
+PWD=`pwd`
```
Comment (john-gress) [inline scripts/buildRpm.sh]: These next four lines repeat lines 7-10.

File: packaging/CodeCoverage.spec (hunk 19 above)
Comment (KjellKod) [inline packaging/CodeCoverage.spec:6]: Can it be that and still be on GitHub.com?

File: packaging/CodeCoverage.spec [hunk 22]
```diff
@@ -0,0 +1,35 @@
+Name:          CodeCoverage
+Version:       %{version}
+Release:       1%{?dist}
+Summary:       A wrapper for building and running code coverage on NetMon C++ repositories
+Group:         Development/Tools
+License:       (c) Copyright 2003-2016, LogRhythm, Inc.  All rights reserved.
```
Comment (john-gress) [inline packaging/CodeCoverage.spec]: @KjellKod is right about this being on a public github account. So we either need to move this to our lrgit repository or change this back to MIT. For now, I'd change it back to MIT.

**Comment by john-gress** [comment]
👍 Be sure to fix the license.

File: scripts/CodeCoverage.py (hunk 17 above)
Comment (craig-cogdill) [inline scripts/CodeCoverage.py]: Addressed

**Comment by craig-cogdill** [comment]
@m1001 Unless otherwise stated, I am deferring your comments (all of which are valid) until the next time we do a Tech Debt Friday. That will allow me to comb through this file with the new knowledge of the python coding guidelines and bring it all up to the standards. A new pull request will be done to address each of the comments that I did not address in this pull request.

I will confer with you when that time comes to make sure everything is up to standard.

**Pull request merged**